README.md
<h3 align="center">
  <br />
  <img src="https://user-images.githubusercontent.com/168240/39561871-51cda852-4e5d-11e8-926b-7692d43143e8.png" alt="logo" width="400" />
  <br />
  <br />
  <br />
</h3>

# cointop

> Coin tracking for hackers

[![License](http://img.shields.io/badge/license-Apache-blue.svg)](https://raw.githubusercontent.com/cointop-sh/cointop/master/LICENSE)
[![Build Status](https://travis-ci.org/cointop-sh/cointop.svg?branch=master)](https://travis-ci.org/cointop-sh/cointop)
[![Go Report Card](https://goreportcard.com/badge/github.com/cointop-sh/cointop?)](https://goreportcard.com/report/github.com/cointop-sh/cointop)
[![GoDoc](https://godoc.org/github.com/cointop-sh/cointop?status.svg)](https://godoc.org/github.com/cointop-sh/cointop)
[![Mentioned in Awesome Terminals](https://awesome.re/mentioned-badge.svg)](https://github.com/k4m4/terminals-are-sexy)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](#contributing)

[`cointop`](https://github.com/cointop-sh/cointop) is a fast and lightweight interactive terminal based UI application for tracking and monitoring cryptocurrency coin stats in real-time.

The interface is inspired by [`htop`](https://en.wikipedia.org/wiki/Htop) and shortcut keys are inspired by [`vim`](https://en.wikipedia.org/wiki/Vim_(text_editor)).

<img src="https://user-images.githubusercontent.com/168240/39569578-7ce9f3b6-4e7a-11e8-82a9-8a18b91b1bd5.png" alt="cointop screenshot" width="880" />

## Demo

This connects to an instance of Cointop using SSH:

```bash
ssh cointop.sh
```

In action

<img src="https://user-images.githubusercontent.com/168240/39569570-75b1547c-4e7a-11e8-8eac-552abaa431f0.gif" alt="screencast" width="880" />

## Table of Contents

Documentation has been moved to [docs.cointop.sh](https://docs.cointop.sh/)

- [Features](#features)
- [Documentation](https://docs.cointop.sh/)
- [Install](https://docs.cointop.sh/install)
- [Update](https://docs.cointop.sh/update)
- [Getting started](https://docs.cointop.sh/getting-started)
- [Shortcuts](https://docs.cointop.sh/shortcuts)
- [Colorschemes](https://docs.cointop.sh/colorschemes)
- [Config](https://docs.cointop.sh/config)
- [SSH server](https://docs.cointop.sh/ssh)
- [FAQ](https://docs.cointop.sh/faq)
- [Contributing](#contributing)
- [Social](#social)
- [Mentioned in](#mentioned-in)
- [Tip Jar](#tip-jar)
- [License](#license)

## Features

- **Shortcut keys**: Vim-inspired shortcut keys, custom key bindings configuration
- **Colorschemes**: Custom colorscheme configuration, 256-color and 24-bit support
- **Favorites**: Save and view favorite coins
- **Portfolio**: Portfolio tracking of holdings, view profit & loss
- **Charts**: Charts for coin price history and global market graphs
- **Search**: Fuzzy searching for finding coins
- **Conversion**: Currency conversion
- **Price Alerts**: Price alerts with desktop notifications
- **Multiple APIs**: Supports multiple coin data APIs; CoinGecko and CoinMarketCap
- **Mouse**: Mouse support
- **Offline**: Offline cache
- **Fast**: Fast sort shortcuts, pagination, chart date range change, auto-refresh
- **Lightweight**: It's very lightweight; can be left running indefinitely

## Contributing

See [docs.cointop.sh/contributing](https://docs.cointop.sh/contributing)

_Many thanks to [Simon Roberts](https://github.com/lyricnz), [Alexis Hildebrandt](https://github.com/afh), and all the [contributors](https://github.com/cointop-sh/cointop/graphs/contributors) that made cointop better._

## Social

- Follow on twitter [@cointop](https://twitter.com/cointop)
- Discuss on [Discord](https://discord.gg/pUVGy7ECGg)

## Mentioned in

Cointop has been mentioned in:

- [@Ubuntu Twitter](https://twitter.com/ubuntu/status/985947962311311360?lang=en)
- [Terminals Are Sexy](https://github.com/k4m4/terminals-are-sexy#tools-and-plugins)
- [The Changelog News](https://changelog.com/news/cointop-coin-tracking-for-hackers-rAzZ)

### Tip Jar

[![BTC Tip Jar](https://img.shields.io/badge/BTC-tip-yellow.svg?logo=bitcoin&style=flat)](https://www.blockchain.com/btc/address/3KdMW53vUMLPEC33xhHAUx4EFtvmXQF8Kf) `3KdMW53vUMLPEC33xhHAUx4EFtvmXQF8Kf`

[![ETH Tip Jar](https://img.shields.io/badge/ETH-tip-blue.svg?logo=ethereum&style=flat)](https://etherscan.io/address/0x9ed3D6793a6b74d8c9A998f5C4b50a25947D53aF) `0x9ed3D6793a6b74d8c9A998f5C4b50a25947D53aF`

Thank you for tips! 🙏

## License

Released under the [Apache 2.0](./LICENSE) license.

© [Miguel Mota](https://github.com/miguelmota)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.6.10] - 2021-11-06
### Added
- Search by symbol or name
- Purchase price option for portfolio entries
- Mouse support for column sorting and menu options

### Changed
- `0` keybinding to go to first row of first page

### Fixed
- Coin sorting
- Editable shortcuts
- Duplicate portfolio entries

## [1.6.9] - 2021-10-12
### Added
- Chart x-axis date labels
- Configurable favorite character
- Configurable chart width
- Save chart height

### Changed
- Renamed organization `miguelmota` → `cointop-sh`

### Fixed
- Global chart currency
- Chart resampling and interpolation
- Chart time periods
- Use preferred cache directory
- Currency symbol width

## [1.6.8] - 2021-09-13
### Fixed
- Hide holdings amount when using command hide flag

## [1.6.7] - 2021-09-13
### Added
- Toggle hide portfolio balances keybinding
- Evaluate expression in portfolio value edit field
- Add 1Y% change column

## [1.6.6] - 2021-08-22
### Added
- Default chart range config

### Fixed
- Duplicate coin portfolio entries
- Increase decimals places shown for small values
- Filecache locking

## [1.6.5] - 2021-04-25
### Added
- Chart fullscreen toggle keybinding
- 24% change to holdings command
- Read environment variables for config

## [1.6.4] - 2021-04-25
### Added
- Preferred cache directory
- Read host numeric monetary locale
- Column filter for holdings command
- SSH server user config type

### Fixed
- Config file path
- String rune count

## [1.6.3] - 2021-03-10
### Added
- Max pages flag
- SSH server connection max timeout

### Fixed
- Negative holdings balance input
- Coins and portfolio row selection
- Table scroll

## [1.6.2] - 2021-02-12
### Added
- Config option to keep row focus on sort

## [1.6.1] - 2021-02-12
### Added
- Multiple coin support in price command

### Fixed
- Chart data interpolation
- CoinMarketCap graph data endpoint

## [1.6.0] - 2021-02-12
### Added
- Configurable table columns
- Basic price alerts

### Fixed
- Coin chart lookup
- Dynamic column widths

## [1.5.5] - 2020-11-15
### Added
- Currency convesion option to holdings command
- Sort by percent holdings shortcut

### Fixed
- Termux cache directory
- Open command on Windows

## [1.5.4] - 2020-08-24
### Added
- Colorschemes directory flag

### Fixed
- Rank order for low market cap coins

## [1.5.3] - 2020-08-14
### Fixed
- Build error

## [1.5.2] - 2020-08-13
### Added
- Holdings command with sorting and filter options
- Bitcoin dominance command

### Fixed
- `XDG_CONFIG_HOME` config path

## [1.5.1] - 2020-08-05
### Fixed
- Version typo

## [1.5.0] - 2020-08-05
### Fixed
- Use version string from go build info

## [1.4.8] - 2020-08-03
### Added
- No cache flag

## [1.4.7] - 2020-08-02
### Added
- SSH server

### Fixed
- Config flag

## [1.4.6] - 2020-05-23
### Fixed
- Decimals places for BTC and ETH currency conversion
- Increase number of page results from CoinGecko

## [1.4.5] - 2020-02-18
### Added
- VND currency conversion

### Fixed
- Convert to chosen currency for market data

## [1.4.4] - 2019-12-31
### Fixed
- Flathub app release version

## [1.4.3] - 2019-12-29
### Added
- Tab keybinding

### Fixed
- Chart update bug fixes
- Marketbar currency bug fixes

## [1.4.2] - 2019-12-29
### Fixed
- Fix keybinding issue on FreeBSD

## [1.4.1] - 2019-11-17
### Fixed
- Fix version ldflags

## [1.4.0] - 2019-11-17
### Added
- Keyboard shortcuts to enlarge and shorten chart

## [1.3.6] - 2019-09-15
### Fixed
- Fixed various navigation and view switching issues

## [1.3.5] - 2019-09-08
### Fixed
- Fixed table sorting issues

## [1.3.4] - 2019-07-05
### Fixed
- Fixed Windows path

## [1.3.3] - 2019-06-30
### Added
- Added price command

## [1.3.2] - 2019-06-30
### Added
- Toggle table fullscreen shortcut and hide view flags

## [1.3.1] - 2019-06-26
### Added
- Show only table option

### Fixed
- CoinGecko prices

## [1.3.0] - 2019-06-09
### Added
- Added colorscheme support

## [1.2.2] - 2019-06-01
### Fixed
- Market bar background color

## [1.2.1] - 2019-06-01
### Fixed
- Added mutex lock when filecaching

## [1.2.0] - 2019-05-12
### Added
- Added CoinGecko API support

### Changed
- Default API from CoinMarketCap to CoinGecko

## [1.1.6] - 2019-04-23
### Added
- Prompt for CoinMarketCap Pro API Key

## [1.1.5] - 2019-04-22
### Fixed
- Release archive to contain latest source code

## [1.1.4] - 2019-04-21
### Added
- Config option to use CoinMarketCap Pro V1 API KEY

### Changed
- CoinMarketCap legacy V2 API to Pro V1 API

## [1.1.3] - 2019-02-25
### Fixed
- Vendor dependencies

## [1.1.2] - 2018-12-30
### Added
- `-clean` flag to clean cache
- `-reset` flag to clean cache and delete config
- `-config` flag to use a different specified config file

### Fixed
- Paginate CoinMarketCap V1 API responses due to their backward-incompatible update

## [1.1.1] - 2018-12-26
### Changed
- Use go modules instead of dep

## [1.1.0] - 2018-12-25
### Added
- Basic portfolio functionality
- `P` keyboard shortcut to toggle portfolio view
- `e` keyboard shortcut to edit portfolio holdings
- `[portfolio]` TOML config holdings list


table.go
package main

import (
	"fmt"
	"strconv"
	"strings"
	"time"

	humanize "github.com/dustin/go-humanize"
	apitypes "github.com/miguelmota/cointop/pkg/api/types"
	"github.com/miguelmota/cointop/pkg/color"
	"github.com/miguelmota/cointop/pkg/pad"
	"github.com/miguelmota/cointop/pkg/table"
)

func (ct *Cointop) updateTable() error {
	maxX, _ := ct.g.Size()
	ct.table = table.New().SetWidth(maxX)
	ct.table.AddCol("")
	ct.table.AddCol("")
	ct.table.AddCol("")
	ct.table.AddCol("")
	ct.table.AddCol("")
	ct.table.AddCol("")
	ct.table.AddCol("")
	ct.table.AddCol("")
	ct.table.AddCol("")
	ct.table.AddCol("")
	ct.table.AddCol("")
	ct.table.AddCol("")
	ct.table.HideColumHeaders = true
	var err error
	if len(ct.coins) == 0 {
		ct.coins, err = ct.fetchData()
		if err != nil {
			return err
		}
	}
	for _, coin := range ct.coins {
		unix, _ := strconv.ParseInt(coin.LastUpdated, 10, 64)
		lastUpdated := time.Unix(unix, 0).Format("15:04:05 Jan 02")
		colorprice := color.Cyan
		color1h := color.White
		color24h := color.White
		color7d := color.White
		if coin.PercentChange1H > 0 {
			color1h = color.Green
		}
		if coin.PercentChange1H < 0 {
			color1h = color.Red
		}
		if coin.PercentChange24H > 0 {
			color24h = color.Green
		}
		if coin.PercentChange24H < 0 {
			color24h = color.Red
		}
		if coin.PercentChange7D > 0 {
			color7d = color.Green
		}
		if coin.PercentChange7D < 0 {
			color7d = color.Red
		}
		ct.table.AddRow(
			pad.Left(fmt.Sprint(coin.Rank), 4, " "),
			pad.Right(coin.Name, 22, " "),
			pad.Right(coin.Symbol, 6, " "),
			colorprice(pad.Left(humanize.Commaf(coin.PriceUSD), 12, " ")),
			pad.Left(humanize.Commaf(coin.MarketCapUSD), 17, " "),
			pad.Left(humanize.Commaf(coin.USD24HVolume), 15, " "),
			color1h(pad.Left(fmt.Sprintf("%.2f%%", coin.PercentChange1H), 9, " ")),
			color24h(pad.Left(fmt.Sprintf("%.2f%%", coin.PercentChange24H), 9, " ")),
			color7d(pad.Left(fmt.Sprintf("%.2f%%", coin.PercentChange7D), 9, " ")),
			pad.Left(humanize.Commaf(coin.TotalSupply), 20, " "),
			pad.Left(humanize.Commaf(coin.AvailableSupply), 18, " "),
			pad.Left(fmt.Sprintf("%s", lastUpdated), 18, " "),
			// add %percent of cap
		)
	}

	ct.table.Format().Fprint(ct.tableview)
	return nil
}

func (ct *Cointop) selectedRowIndex() int {
	_, y := ct.tableview.Origin()
	_, cy := ct.tableview.Cursor()
	return y + cy
}

func (ct *Cointop) selectedCoin() *apitypes.Coin {
	return ct.coins[ct.selectedRowIndex()]
}

func (ct *Cointop) rowLink() string {
	slug := strings.ToLower(strings.Replace(ct.selectedCoin().Name, " ", "-", -1))
	return fmt.Sprintf("https://coinmarketcap.com/currencies/%s", slug)
}
